Nonexistence proofs

SoujanyaPonnapalli opened the pull request:
This PR adds support for the proof of nonexistence.
```
verifyWitness(root, key, witness): Proof of existence
verifyWitness(root, key, witness, false): Proof of non-existence
```

The `merkle-patricia-tree` is traversed based on the key nibbles. Witness contains the tree nodes in the search path of the key. If the node hashes in the witness match up to the root hash and the key nibbles do not match the nibbles in witness; we prove that the key is not present.

A key is not present if the hashes in the witness match up .  to the root and either:
- BranchNode has no branches at the next key nibble or
- If the key nibbles do not match the nibbles in the ExtensionNode or LeafNode

no2chem commented on src/index.ts:
```diff
@@ -1294,25 +1294,34 @@ export class VerificationError extends Error {}
  * check
  * @param key                   A [[Buffer]] containing the key to check
  * @param witness               The [[Witness]] to verify
+ * @param exist                 For existence proofs exist has to be true
```
Is this necessary? The Witness already has the value, so you just need to check Witness.value === null

SoujanyaPonnapalli commented:
> See comments. Also, there doesn't seem to be any tests.

@no2chem Please review the changes
+ Added Tests
+ Fixed the extra args

SoujanyaPonnapalli commented:
verifyWitness(key: Buffer, root: Buffer, witness: RlpWitness).
get(key: Buffer) => witness: Witness.

verifyWitness takes in RlpWitness, but get returns a witness. So for verifying the witness returned by get call, we need to rlpSerialize it. I added it now while adding calls to verifyWitness to verify non-existence proofs.

no2chem commented on src/index.spec.ts:
```diff
@@ -439,11 +448,12 @@ describe('Try batch operations', () => {
     const w1 = tree.get(Buffer.from('a'));
     const w2 = tree.get(Buffer.from('b'));
-    const w3 = tree.get(Buffer.from('c'));
+    const w3 = tree.rlpSerializeWitness(tree.get(Buffer.from('c')));
```
Ok, this is why this is confusing. Why are you putting rlpSerializeWitness here? Just do it below so that the style is consistent.

no2chem commented on src/index.spec.ts:
```diff
@@ -463,6 +473,82 @@ describe('Try batch operations', () => {
 
+describe('verify proofs existence and non-existence', () => {
+  // Keys to create branch, extension and leaf nodes
+  const k1 = Buffer.from('8a40bfaa73256b60764c1bf40675a99083efb075', 'hex');
```
This test is sort of complex. It's good, but it would be nice to have a single unit test that tests for a non-existent node attached to one of each type of node.

SoujanyaPonnapalli commented:
Updated the tests to have a single unit test for testing the non-existence at every node. Also added tests which throw errors if we attempt to verify incorrect proofs
